text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+* **better IO efficiency** (less data to transmit).
+
+![row vs column-oriented](img/0156_OTEL%20-%20Row%20vs%20Column.png)
+
+This OTEP proposes to extend the [OpenTelemetry protocol (OTEP 0035)](https://github.com/open-telemetry/oteps/blob/main/text/0035-opentelemetry-protocol.md)
+with a **generic columnar representation for metrics, logs and traces based on Apache Arrow**. Compared to the existing
+OpenTelemetry protocol this compatible extension has the following improvements:
+
+* **Reduce the bandwidth requirements** of the protocol. The two main levers are: 1) a better representation of the
+  telemetry data based on a columnar representation, 2) a stream-oriented gRPC endpoint that is more efficient to
+  transmit batches of OTLP entities.
+* **Provide a more optimal representation for multivariate time-series data**.
+  With the current version of the OpenTelemetry protocol, users have to transform multivariate time-series (i.e multiple
+  related metrics sharing the same attributes and timestamp) into a collection of univariate time-series resulting in a
+  large amount of duplication and additional overhead covering the entire chain from exporters to backends.
+* **Provide more advanced and efficient telemetry data processing capabilities**. Increasing data volume, cost
+  efficiency, and data minimization require additional data processing capabilities such as data projection,
+  aggregation, and filtering.
+
+These improvements not only address the aforementioned needs but also answer the [open questions](https://github.com/open-telemetry/oteps/blob/main/text/0035-opentelemetry-protocol.md#open-questions)
+cited in OTEP 035 (i.e. cpu usage, memory pressure, compression optimization).
+
+**It is important to understand that this proposal is complementary to the existing protocol. The row-oriented version
+is still suitable for some scenarios. Telemetry sources that generate a small amount of telemetry data should continue
+to use it. On the other side of the spectrum, sources and collectors generating or aggregating a large amount of
+telemetry data will benefit from adopting this extension to optimize the resources involved in the transfer and
+processing of this data. This adoption can be done incrementally.**
+
+Before detailing the specifications of the OTLP Arrow protocol, the following two sections present: 1) a validation of
+the value of a columnar approach based on a set of benchmarks, 2) a discussion of the value of using Apache Arrow as a
+basis for columnar support in OTLP.
+
+### Validation
+
+A series of tests were conducted to compare compression ratios between OTLP and a columnar version of OTLP called OTLP
+Arrow. The two key results are:
+
+* For multivariate time series, OTLP Arrow is **4 times better in terms of bandwidth reduction while improving by
+  3 the speed** of creation + serialization + compression + decompression + deserialization.
```

Comment:
The organization of the data in columns makes it possible to significantly reduce the number of allocations in particular for batches above 500 entities. This has the effect of significantly reducing the cost of memory management (Malloc+GC cost) and therefore better performance overall. This is not true for small batches because the overhead of initializing columns, dictionaries, bitmap validity, and schema becomes more and more important in proportion to the cost of feeding the batch with data. This becomes particularly inefficient for batch sizes below 10, as the cost of initializing the batch becomes greater than the cost of feeding the data. I think that optimizations are still possible in the current code to reduce the minimum size of the batches for which we would still observe a good compression rate and a construction speed equal or inferior to the pure Protobuf approach. For now, I will recommend batches > 500 and when possible 1000 entities.

Reply:
I also updated the OTEP to answer this question, see this [diff](https://github.com/open-telemetry/oteps/pull/171/commits/c5c747caf136c2295bab35d6435a39a1be3ca393).